src/mir/grammar.rkt

```diff
@@ -1,74 +1,65 @@
 #lang racket
-(require redex  "../ty/grammar.rkt")
+(require redex "../decl/grammar.rkt")
 (provide (all-defined-out))
 
-(define-extended-language formality-mir formality-ty
-  (basic-block-map ((basic-block-id basic-block-data) ...))
-  (basic-block-data (statements terminator))
+(define-extended-language formality-mir formality-decl
+  (BasicBlockMap ::= ((BasicBlockId BasicBlockData) ...))
+  (BasicBlockData ::= (Statements Terminator))
 
-  (statements (statement ...))
-  (statement
-   (statement-assign place rvalue)
-   (statement-set-discriminant place variant)
-   (statement-storage-live local-variable-id)
-   (statement-storage-dead local-variable-id)
-   statement-nop
-   )
+  (Statements ::= (Statement ...))
+  (Statement ::=
+             (StatementAssign Place Rvalue)
+             (StatementSetDiscriminant Place VariantId)
+             (StatementStorageLive LocalId)
+             (StatementStorageDead LocalId)
+             StatementNop
+             )
 
-  (rvalue
-   (rvalue-use operand)
-   (rvalue-repeat operand constant)
-   (rvalue-ref lifetime mutability place)
-   (rvalue-addr-of mutability place)
-   (rvalue-len place)
-   (rvalue-binary-op binary-op operand operand)
-   (nullary-op ty)
-   )
+  (Rvalue ::=
+          (RvalueUse Operand)
+          (RvalueRepeat Operand Constant)
+          (RvalueRef Lt MaybeMut Place)
+          (RvalueAddrOf MaybeMut Place)
+          (RvalueLen Place)
+          (RvalueBinaryOp BinaryOp Operand Operand)
+          (RvalueNullaryOp Ty)
+          )
 
-  (binary-op + - * /)
+  (BinaryOp ::= + - * /)
 
-  (lifetime lifetime-id)
+  (Terminator ::=
+              (TerminatorGoto BasicBlockId)
```

Comment:
```suggestion
              (goto BasicBlockId)
```
I think the namespacing isn't necessary since these grammar groups are contextual anyway, and keywords are lowercase

Reply:
I think the namespacing helps in some cases (e.g. `RvalueUse`, which could be confused with use declarations). So, I kept the prefixes for consistency. I'm not sure whether Goto should be lowercase since the `goto` keyword does not actually appear in Rust surface syntax.

Reply:
I believe the capital / lowercase convention is primarily intended to distinguish terminals and nonterminals in redex code. Given the way redex works, I think having such a convention is important, although currently it is not being used uniformly:

* predicates are capital: `WellFormed`, `HasImpl`, `Normalize`
* same for `WhereClause`: `ForAll`, `Implemented`, `Normalize`
* `Ty` constructors are capitalized and sometimes namespaced: `TyRigid`, `TyAlias`, `ForAll` 
* `RigidTy` is a nonterminal, `TyRigid` is a terminal

I wasn't under the impression that all lowercase terminals are supposed to be (derived from) rust surface syntax; there are things like `-outlives-` that don't look anything like rust, and `crate` is not used in the same way as in rust. Personally I would just go for a best effort attempt to look like rust here.

The other reason to avoid namespacing in terminals is that these have to be written out in tests, and we generally want to keep things as short as possible since writing ASTs directly is hard enough as it is. Since these are MIR grammar things, I would suggest we attempt to look like the MIR output produced by rustc where possible, and this uses `goto -> bb1;` for such terminators.

Reply:
Yeah, you've kind of identified the conventions I was shooting for; you've also identified the fact that I've had a certain amount of ambivalence about what is the best naming convention overall.

In general I did try to avoid lowercase words that are not Rust keywords; I adopted `-outlives-` partly because it was not a valid Rust identifier. 

I definitely was not using lowercase as a strict *terminal* identifier, but it kind of evolved a bit by default because I have tried to use Rust keywords for terminals where I can.

Lately I've been wondering about using more symbols, so e.g. we could say `Terminator/Resume` instead of `TerminatorResume`. That said, I think it's probably good to optimize a bit for readaiblity: I'm a bit concerned that the `TerminatorDrop` names etc will make MIR written in PLT Redex form significantly less fun to read. 

I think overall I lean towards: let's use lowercase keywords here, it'll make the MIR nicer to read and I think that's important.